### DISCUSSION TITLE
View-like entity properties

### DISCUSSION THREAD

**Discussion by jczacharia** [mikro-orm/mikro-orm#3533] [Ideas]
This might lean towards a feature request but is there a way to get view-like properties?

For instance something like this:
```ts
@Entity()
export class UserEntity {
  @PrimaryKey()
  id!: string;

  @Property()
  name!: string;

  @Property({ viewable: true })
  async view() {
    return { numOfTodos: await this.todos.loadCount() };
  }

  @OneToMany(() => TodoEntity, (e) => e.user)
  todos = new Collection<TodoEntity>(this);
}

const users = await orm.em.findOne({ id: 'jeremy' });
/*
{
  id: 'jeremy',
  name: 'Jeremy',
  view: {
    numOfTodos: 3
  }
}
*/
```

This can be done currently using by `toJSON` like so:
```ts
const user = await orm.em.findOne({ id: 'jeremy' });
const json = wrap(user).toJSON();
json.view = await json.view;
```
Will produce the result.

I wonder if this could be done automatically by using this `viewable` property option. (Or something of the like)
I think something like this would help a lot of use cases that require an inside look at the count of one-to-many relationships without sending over the populated collection.

**Comment by B4nan** [comment]
You could use virtual properties, which is pretty much what you suggested:

https://mikro-orm.io/docs/defining-entities#virtual-properties

But it cant be async, I dont really see a way to do that, serialization needs to stay sync to support `toJSON` method.

Maybe better for you would be using a formula that would be represented with sql subquery that returns the count:

https://mikro-orm.io/docs/next/defining-entities#formulas

**Reply by jczacharia** [reply]
@B4nan Okay thanks!

I was able to solve it with this:
```ts
  @Formula((alias) => `(SELECT COUNT(*) FROM todo_entity WHERE ${alias}.id = todo_entity.user_id)::int`)
  numOfTodos!: number;
```

In case this helps anyone else.

**Reply by jczacharia** [reply]
Referencing this issue too: https://github.com/mikro-orm/mikro-orm/issues/1599
